mutters-templated-intent/src/main/java/com/rabidgremlin/mutters/templated/TemplatedUtterance.java

```diff
@@ -1,127 +1,178 @@
-  {
-    List<String> input = Arrays.asList(tokenizedUtterance);
-
-    String inputString = StringUtils.join(input, ' ');
-
-    Matcher match = matchPattern.matcher(inputString);
-
-    if (!match.find())
-    {
-      return new TemplatedUtteranceMatch(false);
-    }
-
-    TemplatedUtteranceMatch theMatch = new TemplatedUtteranceMatch(true);
-
-    for (String slotName : slotNames)
-    {
-      Slot slot = slots.getSlot(slotName);
-
-      if (slot == null)
-      {
-        throw new IllegalStateException(
-            "Cannot find slot '" + slotName + " reference by utterace '" + template + "'");
-      }
-
-      String cleanedMatchString = match.group(slotName);
-      List<String> matchedTokens = Arrays.asList(cleanedMatchString.split(" "));
-
-      int matchPos = Collections.indexOfSubList(input, matchedTokens);
-      if (matchPos == -1)
-      {
-        throw new IllegalStateException(
-            "Was unable to find '" + cleanedMatchString + "' in '" + inputString + "'");
-      }
-
-      List<String> orginalTokens = new ArrayList<>();
-      for (int loop = matchPos; loop < matchedTokens.size() + matchPos; loop++)
-      {
-        orginalTokens.add(input.get(loop));
-      }
-
-      SlotMatch slotMatch = slot.match(StringUtils.join(orginalTokens, ' '), context);
-
-      if (slotMatch != null)
-      {
-        theMatch.getSlotMatches().put(slot, slotMatch);
-      }
-      else
-      {
-        return new TemplatedUtteranceMatch(false);
-      }
-    }
-
-    return theMatch;
-  }
-
-  @Override
-  public String toString()
-  {
-    return "Utterance [template=" + template + "]";
-  }
-
-}
+package com.rabidgremlin.mutters.templated;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+import org.apache.commons.lang.StringUtils;
+
+import com.rabidgremlin.mutters.core.Context;
+import com.rabidgremlin.mutters.core.Slot;
+import com.rabidgremlin.mutters.core.SlotMatch;
+import com.rabidgremlin.mutters.core.Slots;
+
+/**
+ * This class handles the matching of an utterance against an utterance
+ * template.
+ * 
+ * It also handles the extraction slot data.
+ * 
+ * @author rabidgremlin
+ * @author dmap
+ *
+ */
+public class TemplatedUtterance
+{
+  private static enum PartType
+  {
+    SIMPLE, SLOT
+  }
+
+  private static class MatchPart
+  {
+    private final PartType type;
+    private final String value;
+
+    public MatchPart(final PartType type, final String value)
+    {
+      this.type = type;
+      this.value = value;
+    }
+
+  }
+
+  private static class SlotMatchRegion
+  {
+    private final String slotName;
+    private final int from, to;
+
+    public SlotMatchRegion(final String slotName, final int from, final int to)
+    {
+      this.slotName = slotName;
+      this.from = from;
+      this.to = to;
+    }
+  }
+
+  private String template;
+
+  private HashSet<String> slotNames = new HashSet<String>();
```

Comment:
Should use the interface **Set**

Reply:
I agree, but I didn't change that line, it was always declared as a HashSet.

If you use the "Hide whitespace changes" setting when viewing the diff you can see the lines I actually changed.  The changed line endings are a pain.